server/rate_counter_test.go

```diff
@@ -0,0 +1,27 @@
+package server
+
+import (
+	"testing"
+	"time"
+)
+
+func TestRateCounter(t *testing.T) {
+	counter := newRateCounter(10)
+
+	var i int
+	for i = 0; i <= 10; i++ {
+		if !counter.allow() {
+			break
+		}
+	}
+
+	if i != 10 {
+		t.Errorf("Expected i = 10, got %d", i)
+	}
+
+	time.Sleep(1100 * time.Millisecond)
```

Comment:
I disagree on this, because that would not properly test `if now.After(r.end) {` condition. The next best thing I can think of is to inject time in to allow:
```
func (r *rateCounter) allow(t time.Time) bool {
```